node/src/exchange.rs

```diff
@@ -83,15 +173,356 @@ impl Codec for HeaderCodec {
+        let mut codec = HeaderCodec {};
+        {
+            let mut reader =
+                ChunkyAsyncRead::<_, 1>::new(Cursor::new(encoded_header_response.clone()));
+            let decoded_header_response = codec
+                .read_response(&stream_protocol, &mut reader)
+                .await
+                .unwrap();
+            assert_eq!(header_response, decoded_header_response[0]);
+        }
+        {
+            let mut reader =
+                ChunkyAsyncRead::<_, 2>::new(Cursor::new(encoded_header_response.clone()));
+            let decoded_header_response = codec
+                .read_response(&stream_protocol, &mut reader)
+                .await
+                .unwrap();
+
+            assert_eq!(header_response, decoded_header_response[0]);
+        }
+        {
+            let mut reader =
+                ChunkyAsyncRead::<_, 3>::new(Cursor::new(encoded_header_response.clone()));
+            let decoded_header_response = codec
+                .read_response(&stream_protocol, &mut reader)
+                .await
+                .unwrap();
+
+            assert_eq!(header_response, decoded_header_response[0]);
+        }
+        {
+            let mut reader =
+                ChunkyAsyncRead::<_, 4>::new(Cursor::new(encoded_header_response.clone()));
+            let decoded_header_response = codec
+                .read_response(&stream_protocol, &mut reader)
+                .await
+                .unwrap();
+
+            assert_eq!(header_response, decoded_header_response[0]);
+        }
+    }
+
+    #[tokio::test]
+    async fn test_chunky_async_read() {
+        let read_data = "FOO123";
+        let cur0 = Cursor::new(read_data);
+        let mut chunky = ChunkyAsyncRead::<_, 3>::new(cur0);
+
+        let mut output_buffer: bytes::BytesMut = b"BAR987".as_ref().into();
+
+        let _ = chunky.read(&mut output_buffer[..]).await.unwrap();
+        assert_eq!(output_buffer, b"FOO987".as_ref());
+    }
+
+    struct ChunkyAsyncRead<T: AsyncRead + Unpin, const CHUNK_SIZE: usize> {
+        inner: T,
+    }
+
+    impl<T: AsyncRead + Unpin, const CHUNK_SIZE: usize> ChunkyAsyncRead<T, CHUNK_SIZE> {
+        fn new(inner: T) -> Self {
+            ChunkyAsyncRead { inner }
+        }
+    }
+
+    impl<T: AsyncRead + Unpin, const CHUNK_SIZE: usize> AsyncRead for ChunkyAsyncRead<T, CHUNK_SIZE> {
+        fn poll_read(
+            mut self: Pin<&mut Self>,
+            cx: &mut Context<'_>,
+            buf: &mut [u8],
+        ) -> Poll<Result<usize, Error>> {
+            let mut small_buf = [0u8; CHUNK_SIZE];
+            match Pin::new(&mut self.inner).poll_read(cx, &mut small_buf) {
+                Poll::Ready(Ok(read)) => {
+                    buf[..read].copy_from_slice(&small_buf[..read]);
+                    Poll::Ready(Ok(read))
+                }
+                ret @ Poll::Ready(Err(_)) => ret,
+                Poll::Pending => Poll::Pending,
+            }
```

Comment:
Intermediate buffer is not needed. You can do this instead:

```suggestion
            let len = buf.len().min(CHUNK_SIZE);
            Pin::new(&mut self.inner).poll_read(cx, &mut buf[..len])
```

Reply:
nice :rocket: 

changed